apps/src/lib/node/ledger/storage/rocksdb.rs

```diff
@@ -1033,35 +1065,22 @@ impl DB for RocksDB {
 
     fn read_merkle_tree_stores(
         &self,
-        height: BlockHeight,
-    ) -> Result<Option<(BlockHeight, MerkleTreeStoresRead)>> {
+        epoch: Epoch,
+        base_height: BlockHeight,
+        store_type: Option<StoreType>,
+    ) -> Result<Option<MerkleTreeStoresRead>> {
         // Get the latest height at which the tree stores were written
         let block_cf = self.get_column_family(BLOCK_CF)?;
-        let height_key = Key::from(height.to_db_key());
-        let key = height_key
-            .push(&"pred_epochs".to_owned())
-            .expect("Cannot obtain a storage key");
-        let pred_epochs: Epochs = match self
-            .0
-            .get_cf(block_cf, key.to_string())
-            .map_err(|e| Error::DBError(e.into_string()))?
-        {
-            Some(b) => types::decode(b).map_err(Error::CodingError)?,
-            None => return Ok(None),
-        };
-        // Read the tree at the first height if no epoch update
-        let stored_height = match pred_epochs.get_epoch_start_height(height) {
-            Some(BlockHeight(0)) | None => BlockHeight(1),
-            Some(h) => h,
-        };
-
-        let tree_key = Key::from(stored_height.to_db_key())
-            .push(&"tree".to_owned())
-            .map_err(Error::KeyError)?;
         let mut merkle_tree_stores = MerkleTreeStoresRead::default();
-        for st in StoreType::iter() {
-            let prefix_key =
-                tree_key.push(&st.to_string()).map_err(Error::KeyError)?;
+        let store_types = store_type
+            .map(|st| vec![st])
+            .unwrap_or(StoreType::iter().copied().collect());
```

Comment:
[`Either`](https://docs.rs/itertools/latest/itertools/enum.Either.html#impl-Iterator-for-Either%3CL,+R%3E) from itertools implements `Iterator`, so you don't need to allocate here. not really a big deal I guess, more like a FYI